<a href="https://colab.research.google.com/github/keerthigoud1536/Handling-Imbalanced-Classification-Using-Sampling-Techniques-Machine-Learning/blob/main/Imbalanced_Dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Imbalanced Dataset Project**


## **STEP 1 — Import Required Libraries**

In [1]:
# Basic libraries
import numpy as np
import pandas as pd



## **STEP 2 — Read Dataset**

In [2]:
#load the dataset
df_train = pd.read_csv("/content/train.csv")
df_train.head()

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
0   7       0          2              2          5              1   
1   9       0          1              1          7              0   
2  13       0          5              4          9              1   
3  16       0          0              1          2              0   
4  17       0          0              2          0              1   

   ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  ...  \
0              0              0              1              0  ...   
1              0              0              0              1  ...   
2              0              0              0              1  ...   
3              0              1              0              0  ...   
4              0              1              0              0  ...   

   ps_calc_11  ps_calc_12  ps_calc_13  ps_calc_14  ps_calc_15_bin  \
0           9           1           5           8               0   
1           3           1           1           9               0   
2           4           2           7           7               0   
3           2           2           4           9               0   
4           3           1           1           3               0   

   ps_calc_16_bin  ps_calc_17_bin  ps_calc_18_bin  ps_calc_19_bin  \
0               1               1               0               0   
1               1               1               0               1   
2               1               1               0               1   
3               0               0               0               0   
4               0               0               1               1   

   ps_calc_20_bin  
0               1  
1               0  
2               0  
3               0  
4               0  

[5 rows x 59 columns]

## **STEP 3 — EDA (Exploratory Data Analysis)**

In [3]:
print(df_train.shape)

(595212, 59)


In [4]:
print(df_train.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 595212 entries, 0 to 595211
Data columns (total 59 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   id              595212 non-null  int64  
 1   target          595212 non-null  int64  
 2   ps_ind_01       595212 non-null  int64  
 3   ps_ind_02_cat   595212 non-null  int64  
 4   ps_ind_03       595212 non-null  int64  
 5   ps_ind_04_cat   595212 non-null  int64  
 6   ps_ind_05_cat   595212 non-null  int64  
 7   ps_ind_06_bin   595212 non-null  int64  
 8   ps_ind_07_bin   595212 non-null  int64  
 9   ps_ind_08_bin   595212 non-null  int64  
 10  ps_ind_09_bin   595212 non-null  int64  
 11  ps_ind_10_bin   595212 non-null  int64  
 12  ps_ind_11_bin   595212 non-null  int64  
 13  ps_ind_12_bin   595212 non-null  int64  
 14  ps_ind_13_bin   595212 non-null  int64  
 15  ps_ind_14       595212 non-null  int64  
 16  ps_ind_15       595212 non-null  int64  
 17  ps_ind_16_

In [5]:
print(df_train.describe())

                 id         target      ps_ind_01  ps_ind_02_cat  \
count  5.952120e+05  595212.000000  595212.000000  595212.000000   
mean   7.438036e+05       0.036448       1.900378       1.358943   
std    4.293678e+05       0.187401       1.983789       0.664594   
min    7.000000e+00       0.000000       0.000000      -1.000000   
25%    3.719915e+05       0.000000       0.000000       1.000000   
50%    7.435475e+05       0.000000       1.000000       1.000000   
75%    1.115549e+06       0.000000       3.000000       2.000000   
max    1.488027e+06       1.000000       7.000000       4.000000   

           ps_ind_03  ps_ind_04_cat  ps_ind_05_cat  ps_ind_06_bin  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        4.423318       0.416794       0.405188       0.393742   
std         2.699902       0.493311       1.350642       0.488579   
min         0.000000      -1.000000      -1.000000       0.000000   
25%         2.000000       0.000000       

In [6]:
print(df_train.isnull().sum())

id                0
target            0
ps_ind_01         0
ps_ind_02_cat     0
ps_ind_03         0
ps_ind_04_cat     0
ps_ind_05_cat     0
ps_ind_06_bin     0
ps_ind_07_bin     0
ps_ind_08_bin     0
ps_ind_09_bin     0
ps_ind_10_bin     0
ps_ind_11_bin     0
ps_ind_12_bin     0
ps_ind_13_bin     0
ps_ind_14         0
ps_ind_15         0
ps_ind_16_bin     0
ps_ind_17_bin     0
ps_ind_18_bin     0
ps_reg_01         0
ps_reg_02         0
ps_reg_03         0
ps_car_01_cat     0
ps_car_02_cat     0
ps_car_03_cat     0
ps_car_04_cat     0
ps_car_05_cat     0
ps_car_06_cat     0
ps_car_07_cat     0
ps_car_08_cat     0
ps_car_09_cat     0
ps_car_10_cat     0
ps_car_11_cat     0
ps_car_11         0
ps_car_12         0
ps_car_13         0
ps_car_14         0
ps_car_15         0
ps_calc_01        0
ps_calc_02        0
ps_calc_03        0
ps_calc_04        0
ps_calc_05        0
ps_calc_06        0
ps_calc_07        0
ps_calc_08        0
ps_calc_09        0
ps_calc_10        0
ps_calc_11        0


**Proportion Of Class**

In [7]:
target_count = df_train.target.value_counts()
print('Class 0:', target_count[0])
print('Class 1:', target_count[1])
print('Proportion of class 0 is ', round(target_count[0] * 100 / (target_count[1] + target_count[0]), 2),'%')

Class 0: 573518
Class 1: 21694
Proportion of class 0 is  96.36 %


## **STEP 4 — Train-Test Split**

**Split Features &Target**

In [8]:
X = df_train.drop("target",axis=1)
y = df_train["target"]

In [9]:
X.head()
y.head()

0    0
1    0
2    0
3    0
4    0
Name: target, dtype: int64

In [10]:
!pip install imblearn

**Splitting The Data**

In [11]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

**Scaling The Data**

In [12]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)



##  **Step 4: Apply Cluster Centroids Under-Sampling**



In [18]:
from imblearn.under_sampling import ClusterCentroids
cc = ClusterCentroids(sampling_strategy={0:100},random_state=42)
X_cc, y_cc = cc.fit_resample(X_train, y_train)

print("After Under-sampling:")
print(y_cc.value_counts())

After Under-sampling:
target
1    17355
0      100
Name: count, dtype: int64


## **STEP 6 — ML Models on Undersampled Data**

In [19]:
# ML Models
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
# Evaluation metrics
from sklearn.metrics import f1_score, roc_auc_score

models = {
    "Logistic Regression": LogisticRegression(max_iter=300),
    "Decision Tree": DecisionTreeClassifier()
}

print("\n===== UNDERSAMPLING RESULTS ====")
for name, model in models.items():
    model.fit(X_cc, y_cc)
    y_pred = model.predict(X_test)
    y_prob = model.predict_proba(X_test)[:, 1]

    print("\nModel:", name)
    print("F1 Score:", f1_score(y_test, y_pred))
    print("AUC Score:", roc_auc_score(y_test, y_prob))



===== UNDERSAMPLING RESULTS ====


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(



Model: Logistic Regression
F1 Score: 0.07034499533790084
AUC Score: 0.5252200511465672

Model: Decision Tree
F1 Score: 0.07033440858472062
AUC Score: 0.5


## **STEP 7 — Random Oversampling**

In [20]:
from imblearn.over_sampling import RandomOverSampler
ros = RandomOverSampler()
X_ros, y_ros = ros.fit_resample(X_train, y_train)

print("After Random Oversampling:")
print(y_ros.value_counts())


After Random Oversampling:
target
0    458814
1    458814
Name: count, dtype: int64


## **STEP 8 — SMOTE Oversampling**

In [21]:
from imblearn.over_sampling import SMOTE
sm = SMOTE()
X_sm, y_sm = sm.fit_resample(X_train, y_train)

print("After SMOTE:")
print(y_sm.value_counts())


After SMOTE:
target
0    458814
1    458814
Name: count, dtype: int64


## **STEP 9 — ML Models on SMOTE Data**

In [22]:
# ML Models
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
# Evaluation metrics
from sklearn.metrics import f1_score, roc_auc_score

models = {
    "Logistic Regression": LogisticRegression(max_iter=300),
    "Decision Tree": DecisionTreeClassifier()
}

print("\n===== SMOTE RESULTS =====")
for name, model in models.items():
    model.fit(X_sm, y_sm)
    y_pred = model.predict(X_test)
    y_prob = model.predict_proba(X_test)[:,1]

    print("\nModel:", name, "(SMOTE)")
    print("F1 Score:", f1_score(y_test, y_pred))
    print("AUC Score:", roc_auc_score(y_test, y_prob))



===== SMOTE RESULTS =====


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(



Model: Logistic Regression (SMOTE)
F1 Score: 0.0748641441927971
AUC Score: 0.5411019972615829

Model: Decision Tree (SMOTE)
F1 Score: 0.058975685463010866
AUC Score: 0.5092584818694714


# **REPORT**

**UnderSampling using Cluster Centroids**

Model: Logistic Regression
F1 Score: 0.07034499533790084
AUC Score: 0.5252200511465672

Model: Decision Tree
F1 Score: 0.07033440858472062
AUC Score: 0.5

**Over Sampling Using SMOTE**
Model: Logistic Regression (SMOTE)
F1 Score: 0.0748641441927971
AUC Score: 0.5411019972615829

Model: Decision Tree (SMOTE)
F1 Score: 0.058975685463010866
AUC Score: 0.5092584818694714

# **Final Conclusion**

-SMOTE + Logistic Regression / Decision Tree usually gives the best balanced performance.

-Cluster Centroids is useful but SMOTE performed better overall.